### PR TITLE
fix(codecs): Fix framing handling of long frames

### PR DESCRIPTION
**Pull request by jszwedko**
Previously it would return `Ok(None)` when a frame longer than
`max_length` was encountered, but the `tokio_codec::Decoder#decode` trait method specifies:

> If an entire frame is available, then this instance will remove those bytes from the buffer provided and return them as a decoded frame. Note that removing bytes from the provided buffer doesn't always necessarily copy the bytes, so this should be an efficient operation in most circumstances.
>
> If the bytes look valid, but a frame isn't fully available yet, then Ok(None) is returned. This indicates to the Framed instance that it needs to read some more bytes before calling this method again.
>
> Note that the bytes provided may be empty. If a previous call to decode consumed all the bytes in the buffer then decode will be called again until it returns Ok(None), indicating that more bytes need to be read.

That is, it should only return `Ok(None)` if it hit the end of the
buffer without reading a full frame.

As a reference, consider the implementation of `tokio_codec::LinesCodec`
which, when discarding, discards until it can find the next frame before
returning (granted it behaves slightly differently by returning an
error, first, when it hits a line that is too long where we just warn).

Signed-off-by: Jesse Szwedko <jesse@szwedko.me>

### CODE REVIEW

**Review by blt — APPROVED** [review]

**Comment by pablosichert** [comment]
Oh damn, that is quite a catch. If I recall correctly, I copied this implementation over, so that bug must have existed quite some time already.

File: src/codecs/framing/character_delimited.rs [hunk 1]
```diff
@@ -214,13 +213,20 @@ mod tests {
...
-        assert!(codec.decode(buf).unwrap().is_none());
-        assert!(codec.decode(buf).unwrap().is_some());
-        assert!(codec.decode_eof(buf).unwrap().is_none());
-        assert!(codec.decode_eof(buf).unwrap().is_some());
+        assert_eq!(codec.decode(buf).unwrap(), Some(Bytes::from("123456")));
+        assert_eq!(codec.decode(buf).unwrap(), None);
+
+        let buf = &mut BytesMut::new();
+
+        buf.reserve(200);
```
Comment (pablosichert) [inline src/codecs/framing/character_delimited.rs:224]: That seems a bit arbitrary / unnecessary in this test. Or am I missing something?

File: src/codecs/framing/newline_delimited.rs [hunk 2]
```diff
@@ -134,9 +134,7 @@ mod tests {
         let mut decoder = NewlineDelimitedDecoder::new_with_max_length(3);
 
         assert_eq!(decoder.decode(&mut input).unwrap().unwrap(), "foo");
-        assert_eq!(decoder.decode(&mut input).unwrap(), None);
         assert_eq!(decoder.decode(&mut input).unwrap().unwrap(), "baz");
-        assert_eq!(decoder.decode(&mut input).unwrap(), None);
```
Comment (pablosichert) [inline src/codecs/framing/newline_delimited.rs:139]: This line should still hold, right? I'd leave it in in that case.

File: src/codecs/framing/character_delimited.rs (hunk 1 above)
Comment (jszwedko) [reply src/codecs/framing/character_delimited.rs:224]: Yeah, I'm not really sure why we are reserving. That already existed in the tests, but I'll drop it in both places.

File: src/codecs/framing/character_delimited.rs [hunk 3]
```diff
@@ -214,13 +213,20 @@ mod tests {
...
+        // limit is 6 so it will skip longer lines
+        buf.put_slice(b"1234567\n123456\n123412314\n123");
+
+        assert_eq!(codec.decode_eof(buf).unwrap(), Some(Bytes::from("123456")));
+        assert_eq!(codec.decode_eof(buf).unwrap(), Some(Bytes::from("123")));
```
Comment (pablosichert) [inline src/codecs/framing/character_delimited.rs:229]: Would also add a last line of

```rust
assert_eq!(decoder.decode_eof(buf).unwrap(), None);
```

for good measure

File: src/codecs/framing/newline_delimited.rs (hunk 2 above)
Comment (jszwedko) [reply src/codecs/framing/newline_delimited.rs:139]: Good call 👍

**Comment by blt** [comment]
I'll wait until this goes up before addressing #10569

**Review by pablosichert — APPROVED** [review]
Great catch @jszwedko!

This is quite a frightening bug. Glad we caught it, but could need some follow-up to how we let that one slip.

**Pull request merged**